gradio/processing_utils.py

```diff
@@ -273,121 +274,160 @@ def save_file_to_cache(file_path: str | Path, cache_dir: str) -> str:
+    current_url = url
 
-            for chunk in response.iter_raw():
-                f.write(chunk)
+    async with httpx.AsyncClient() as client:
+        while redirect_count < max_redirects:
+            try:
+                await validate_url(current_url)
 
-    return full_temp_file_path
+                async with client.stream(
+                    "GET", current_url, follow_redirects=False
+                ) as response:
+                    if response.is_redirect:
+                        redirect_count += 1
+                        current_url = response.headers["Location"]
+                        continue
 
+                    content_disposition = response.headers.get("Content-Disposition")
+                    if content_disposition and "filename=" in content_disposition:
+                        filename = Path(
+                            content_disposition.split("filename=")[1].strip('"')
+                        ).name
 
-async def async_save_url_to_cache(url: str, cache_dir: str) -> str:
-    """Downloads a file and makes a temporary file path for a copy if does not already
-    exist. Otherwise returns the path to the existing temp file. Uses async httpx."""
-    url = get_public_url(url)
+                    else:
+                        filename = client_utils.strip_invalid_filename_characters(
+                            Path(url).name
+                        )
 
-    temp_dir = hash_url(url)
-    temp_dir = Path(cache_dir) / temp_dir
-    temp_dir.mkdir(exist_ok=True, parents=True)
-    name = client_utils.strip_invalid_filename_characters(Path(url).name)
-    full_temp_file_path = str(abspath(temp_dir / name))
+                    filepath = os.path.join(cache_dir, filename)
 
-    if not Path(full_temp_file_path).exists():
-        async with async_client.stream("GET", url, follow_redirects=True) as response:
-            for redirect in response.history:
-                get_public_url(str(redirect.url))
+                    async with asyncio.Lock():
+                        with open(filepath, "wb") as f:
+                            async for chunk in response.aiter_bytes():
+                                f.write(chunk)
 
-            async with aiofiles.open(full_temp_file_path, "wb") as f:
-                async for chunk in response.aiter_raw():
-                    await f.write(chunk)
+                    return filepath
 
-    return full_temp_file_path
+            except httpx.HTTPStatusError as e:
+                raise ValueError(
+                    f"HTTP error occurred requesting resource {url}: {e}"
+                ) from e
+            except Exception as e:
+                raise ValueError(
+                    f"An error occurred requesting resource {url}: {e}"
+                ) from e
+
+    raise ValueError("Too many redirects")
+
+
+def sync_ssrf_protected_httpx_download(url: str, cache_dir: str) -> str:
+    return asyncio.run(ssrf_protected_httpx_download(url, cache_dir))
```

Comment:
We shouldn't use asyncio.run here because it will error if there is already an event loop running.

Demo to test:

```python
import gradio as gr

def return_video_url():
    return "https://cdn.pixabay.com/video/2024/07/14/221180_tiny.mp4"

with gr.Blocks() as demo:
    vid = gr.Video(format="ogg")
    gen = gr.Button("Generate Video")
    gen.click(return_video_url, None, vid)

demo.launch()
```

Reply:
We should use `anyio.to_thread.run_sync`? Or check if there is a current event loop running ?

Reply:
Good catch, just handled the case an event loop is already running. Though `anyio.to_thread.run_sync ` is for running sync stuff in async contexts, we're doing the opposite